Tool calling support. Part I

rohitprasad15 opened the pull request:
Add tool calling support for below providers -
OpenAI, Groq, Anthropic, AWS, & Mistral.
OpenAI compatible SDKs need to changes for
tool calling support.

Adding utility ToolManager for users to easily
supply tools, and parse model's request for
tool usage.

ksolo commented on aisuite/providers/anthropic_provider.py:
```diff
@@ -15,26 +34,195 @@ def __init__(self, **config):
+
+    def _convert_single_message(self, msg):
+        """Convert a single message to Anthropic format."""
+        if isinstance(msg, dict):
+            return self._convert_dict_message(msg)
+        return self._convert_message_object(msg)
```
It looks like both `_convert_dict_message` and `_convert_message_object` do the same things. Would it be better to convert to one format and keep one of the convert message methods.

```python
if isinstance(msg, dict):
    msg = Message(msg)
return self._convert_message_object(msg)
```

or the other way
```python
if isinstance(msg, Message):
    msg = dict(msg)
return self._convert_dict_message(msg)
```

ksolo commented on aisuite/providers/anthropic_provider.py:
```diff
@@ -15,26 +34,195 @@ def __init__(self, **config):
+            message_content.append({"type": "text", "text": content})
+
+        for tool_call in tool_calls:
+            tool_input = (
+                tool_call["function"]["arguments"]
+                if isinstance(tool_call, dict)
+                else tool_call.function.arguments
+            )
+            message_content.append(
+                {
+                    "type": "tool_use",
+                    "id": (
+                        tool_call["id"] if isinstance(tool_call, dict) else tool_call.id
+                    ),
+                    "name": (
+                        tool_call["function"]["name"]
+                        if isinstance(tool_call, dict)
+                        else tool_call.function.name
+                    ),
+                    "input": json.loads(tool_input),
+                }
+            )
```
Similar thought, converting to either an object or dict at the beginning of this would create a consistency that would not require the `if/else` statements throughout the function call.

ksolo commented on aisuite/utils/tool_manager.py:
Can you add unit tests for this?

ksolo reviewed (approved):
Looks good @rohitprasad15 

Nothing blocking, just a few suggestions in the review.

mmtmn reviewed (approved):
outside of unit tests, lgtm

guilherme-vargas commented:
Could i help on this adding support for GoogleProvider ?

amaarora commented:
@rohitprasad15 When can we expect tool calling in `aisuite`? Is there a timeline?

rohitprasad15 commented:
@ksolo  - PTAL again. Added support for a bunch of other providers, and added test cases.
I will take care of the Lint and other failures before merging. Thanks!
